Moves samples from ember-samples/hot-reloaded-app to the dummy app so…

MiguelMadero opened the pull request:
… it’s all better self-contained

All the following examples are in dummy and working: 

```
{{mixed-classic}}
{{mixed-pod}}
{{inline-template-classic}}
{{inline-template-pod}}
{{js-only-classic}}
{{js-only-pod}}
{{template-only-classic}}
{{template-only-pod}}
```

See: 

![livereload](https://cloud.githubusercontent.com/assets/47388/18810507/c30ea424-826e-11e6-8485-7dfcfc5f317c.gif)

MiguelMadero commented on addon/mixins/hot-component.js:
```diff
@@ -5,10 +5,12 @@ export default Ember.Mixin.create({
   init () {
```
We used to have two classes inheriting from this one at some point, now that there is only one I'll inline the mixin.

toranb replied:
I like it - not a fan of using the mixin these days unless is literally the only option so I'm glad to see this go away

MiguelMadero replied:
Will remove on a separate PR

MiguelMadero commented on addon/services/hot-reload.js:
```diff
@@ -1,9 +1,62 @@
-export default Ember.Service.extend(Ember.Evented, {
-  init () {
-    //TODO: non global alternatives please :)
-    window.hotReloadService = this;
```
This was only used for debugging purposes.

toranb replied:
Oh, I assumed the event handling (from JS to ember) required it + the dev tools plugin. That not the case ?

MiguelMadero replied:
No. that was just me playing in the console.

MiguelMadero commented on addon/instance-initializers/hot-loader-livereload-plugin.js:
```diff
@@ -7,26 +7,31 @@ function createPlugin (appName, hotReloadService) {
-    for (var i = tags.length; i >= 0; i--){
-      if (tags[i] && tags[i].getAttribute('src') != null && tags[i].getAttribute('src').indexOf(appName) !== -1) {
-        tags[i].parentNode.removeChild(tags[i]);
+    const cancelableEvent = { modulePath: path, cancel: false};
```
I'm not super convinced this is the best pattern. Any suggestions?

toranb replied:
I'm open to anything that's better :)

I mostly followed what the livereload did for css (removing/then adding the link) but if that isn't needed/or desired for the JS side I'm cool with whatever (no emotional attachment to any of this code)

MiguelMadero replied:
I think th script tag pattern is fine. The one im not sure about is the cancelable event.

MiguelMadero commented on addon/instance-initializers/hot-loader-livereload-plugin.js:
```diff
@@ -7,26 +7,31 @@ function createPlugin (appName, hotReloadService) {
+    hotReloadService.trigger('willLiveReload', cancelableEvent);
+    if (cancelableEvent.cancel) {   // Only hotreload if someone canceled the regular reload
+      // TODO: is this needed? Consider removing since it might have undersired effects
+      window.runningTests = true;
```
@toranb do you know if this is needed and why?

toranb replied:
so if you don't do the runningTests = true the application will fire the "create" again (Ember.Application.create)

MiguelMadero replied:
Oh. Good to know. Ill add a comment about that since it isnt obvious. 

- [ ] document the use of window.testing

MiguelMadero commented on addon/instance-initializers/hot-loader-livereload-plugin.js:
```diff
@@ -7,26 +7,31 @@ function createPlugin (appName, hotReloadService) {
+      script.onload = function() {
+        setTimeout(function() {
+          window.runningTests = false;
+          hotReloadService.trigger('newChanges', path);
```
It might just be the event names what I don't like. Any suggestions? Maybe `willLiveReload` and `willHotReload` or `willLiveReload` and `liveReloadAborted`...

toranb replied:
agreed - newChanges is fairly v1 generic :)

I'm cool with willHotReload and willLiveReload

MiguelMadero replied:
Cool. Will rename once i get to my computer. 

- [ ] rename

MiguelMadero commented on addon/instance-initializers/hot-loader-livereload-plugin.js:
```diff
@@ -7,26 +7,31 @@ function createPlugin (appName, hotReloadService) {
+        }, 10);
+      };
+      script.type = 'text/javascript';
+      script.src = `/assets/${appName}.js`;
```
minor/later: this will break engines... I'll deal with that later, I have a few ideas and I'll need it at some point.

toranb replied:
ah good point - yeah I imagine the official engines support could be a v2 feature (it's still not in ember core as of 2.8 anyway as you need to bolt on last I checked)

MiguelMadero replied:
Afaik engines is supported in 2.8. All the internals are there, only lazy load is experimental which is the part thst changes the build. Def v2 for us

MiguelMadero commented on addon/instance-initializers/hot-loader-livereload-plugin.js:
```diff
@@ -41,7 +46,12 @@ export function initialize(appInstance) {
   }
-  const Plugin = createPlugin(appInstance.base.name, appInstance.lookup('service:hot-reload'));
+  let appName = appInstance.base.name;
+  if (appName === 'ember-cli-hot-loader') {
```
We need a better way to detect if we're running an addon's dummy app, for now, this only works with apps and with this addon. I created https://github.com/toranb/ember-cli-hot-loader/issues/20 as a reminder to fix it.

toranb replied:
Great catch - this wasn't even on my radar

toranb commented:
@MiguelMadero amazing progress here! That animated gif above built using the code/examples from this PR? They all working in 2x (including pure hbs files/ mixed hbs + js and layout (in JS only)) ?? If so - you are absolutely crushing it man :) #keep_up_the_great_work

MiguelMadero commented:
Yes. All of those work in 2x (2.7). I got those working in 1.13, but Ill
send a separare PR for that.
On Sat, Sep 24, 2016 at 9:42 PM Toran Billups <notifications@github.com>
wrote:

> @MiguelMadero <https://github.com/MiguelMadero> amazing progress here!
> That animated gif above built using the code/examples from this PR? They
> all working in 2x (including pure hbs files/ mixed hbs + js and layout (in
> JS only)) ?? If so - you are absolutely crushing it man :)
> #keep_up_the_great_work
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/toranb/ember-cli-hot-loader/pull/17#issuecomment-249395684>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC5HD5Sq66CcEtqF6JIChXT4587vIjsks5qtcOHgaJpZM4JvkEC>
> .
>

MiguelMadero commented:
I just pushed the fixes for Ember 1.13. This PR will be easier to review commit by commit, each commit fixes a different issue.

toranb commented:
@MiguelMadero this looks amazing - any reason we shouldn't merge this and get people using it ?

MiguelMadero commented:
All good. Ill send a different PR with the tests and some inline comments. Ill keep an eye and can release a 0.0.3 once this makes it to master.

MiguelMadero commented:
@toranb I pushed a few new changes, I added tests for clear-cache and addressed some of the earlier tasks on this PR. Since I renamed the event newChanges to willHotReload I think it makes sense to releases as 0.1.0 instead of 0.0.3 since it's a pre-release breaking change to a public API. Not too concerned about backwards compatibility since I don't expect anyone outside of the addon relying on it.

MiguelMadero commented on .travis.yml:
```diff
@@ -19,6 +19,8 @@ env:
   fast_finish: true
   allow_failures:
+    # For now it doesn't work in Glimmer 2. We'll try to fix before the release.
+    - env: EMBER_TRY_SCENARIO=ember-beta
```
We knew this doesn't work, but we didn't have any tests for it. Now at least we know it breaks. Unfortunately, it breaks in more places than what was I expecting.

toranb replied:
@MiguelMadero should we go ahead and remove this now that v0.1.3 is out? I would assume so -that way we could find out it breaks/ or works when ember v2.11.0-beta starts shipping soon :)

MiguelMadero commented:
Great idea. Do you want to go for it?

On Sat, Nov 19, 2016 at 4:24 PM Toran Billups <notifications@github.com>
wrote:

> *@toranb* commented on this pull request.
> ------------------------------
>
> In .travis.yml <https://github.com/toranb/ember-cli-hot-loader/pull/17>:
>
> > @@ -19,6 +19,8 @@ env:
>  matrix:
>    fast_finish: true
>    allow_failures:
> +    # For now it doesn't work in Glimmer 2. We'll try to fix before the release.
> +    - env: EMBER_TRY_SCENARIO=ember-beta
>
> @MiguelMadero <https://github.com/MiguelMadero> should we go ahead and
> remove this now that v0.1.3 is out? I would assume so -that way we could
> find out it breaks/ or works when ember v2.11.0-beta starts shipping soon :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/toranb/ember-cli-hot-loader/pull/17>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAC5HNxHlwoWgd7NKyrPnzMm8YmihDFHks5q_0z3gaJpZM4JvkEC>
> .
>

toranb commented:
@MiguelMadero yessir - I'll get a quick PR up to include this going fwd
